### PR TITLE
HDInsight and Kubernetes Require Subscription  Subnet

### PR DESCRIPTION
**Pull request by evsmithmsft**
Two policies to requrie VNet Injection for new HDInsight and Kubernetes clusters

Recommended Network Isolation Policies

### CODE REVIEW

File: samples/Network/hdinsight-require-subscription-vnet/azurepolicy.json [hunk 1]
```diff
@@ -0,0 +1,56 @@
+{
+    "mode": "all",
```
Comment (camillemarie) [inline samples/Network/hdinsight-require-subscription-vnet/azurepolicy.json]: >all [](start = 13, length = 3)

"indexed"

File: samples/Network/hdinsight-require-subscription-vnet/azurepolicy.json [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    "policyRule": {
+      "if": {
+        "anyOf": [
+          {
+            "allOf": [
+              {
+                "field": "type",
+                "equals": "Microsoft.HDInsight/clusters"
+              },
```
Comment (camillemarie) [inline samples/Network/hdinsight-require-subscription-vnet/azurepolicy.json]: This type check is repeated.

Could be simplified by having the "allOf" on the outside, with the "type" condition first, then "anyOf" inside the second condition

Comment (camillemarie) [reply samples/Network/hdinsight-require-subscription-vnet/azurepolicy.json]: `
     "allOf": [
        {
          "field": "type",
          "equals": "Microsoft.HDInsight/clusters"
        },
        {
          "anyOf": [
            {
              "not": {
                "field": "Microsoft.HDInsight/clusters/computeProfile.roles[*].virtualNetworkProfile.subnet",
                "notEquals": ""
              }
            },
            {
              "field": "Microsoft.HDInsight/clusters/computeProfile.roles[*].virtualNetworkProfile.subnet",
              "exists": "false"
            }
          ]
        }
      ]`

---
In reply to: [268248473](https://github.com/Azure/azure-policy/pull/340#discussion_r268248473) [](ancestors = 268248473)

**Comment by evsmithmsft** [comment]
Yes, I noticed that when I uploaded, don't remember why I did it the other way around when I originally wrote it last year.  I suspect the code morphed from something else :).  Both the policies are the same, I can switch and resubmit

File: samples/Network/hdinsight-require-subscription-vnet/azurepolicy.parameters.json [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    "metadata": {
+        "displayName": "Effect",
+        "description": "Enable or disable the execution of the policy"
+    },
+    "allowedValues": [
+        "Audit",
+        "Deny",
+        "Disabled"
+    ],
+    "defaultValue": "Audit"
```
Comment (camillemarie) [inline samples/Network/hdinsight-require-subscription-vnet/azurepolicy.parameters.json]: nit: Please indent consistently
(your favorite text editor can probably do this for you :])

File: samples/Network/hdinsight-require-subscription-vnet/README.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+## Try on Portal
+
+[![Deploy to Azure](http://azuredeploy.net/deploybutton.png)](https://portal.azure.com/#blade/Microsoft_Azure_Policy/CreatePolicyDefinitionBlade/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-policy%2Fmaster%2Fsamples%2FNetwork%2Fhdinsight-require-subscription-vnet%2Fazurepolicy.json)
+
+## Try with PowerShell
+
+````powershell
+$definition = New-AzPolicyDefinition -Name "deploy-network-watcher-when-virtual-network-created" -DisplayName "Deploy HDInsight Cluster with Subscription's VNet" -description "This policy enforces use of subscription's virtual network when creating HDInsight Clusters, allowing securing network level access to the HDInsight Cluster." -Policy 'https://raw.githubusercontent.com/Azure/azure-policy/master/samples/Network/hdinsight-require-subscription-vnet/azurepolicy.rules.json' -Parameter 'https://raw.githubusercontent.com/Azure/azure-policy/master/samples/Network/hdinsight-require-subscription-vnet/azurepolicy.parameters.json' -Mode Indexed
```
Comment (camillemarie) [inline samples/Network/hdinsight-require-subscription-vnet/README.md]: >deploy-network-watcher-when-virtual-network-created" [](start = 44, length = 52)

update this to hdinsight title (not network watcher)

File: samples/Network/hdinsight-require-subscription-vnet/azurepolicy.json [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+              {
+                "not": {
+                  "field": "Microsoft.HDInsight/clusters/computeProfile.roles[*].virtualNetworkProfile.subnet",
+                  "notIn": [
+                    "null",
```
Comment (camillemarie) [inline samples/Network/hdinsight-require-subscription-vnet/azurepolicy.json]: >null [](start = 21, length = 4)

Have you seen "null" (as a string) in your testing? 

This is a new one to me.

File: samples/Network/kubernetes-require-subscription-vnet/azurepolicy.json [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    "mode": "all",
+    "policyRule": {
+      "if": {
+        "anyOf": [
+          {
+            "allOf": [
+              {
+                "field": "type",
+                "equals": "Microsoft.ContainerService/managedClusters"
+              },
```
Comment (camillemarie) [inline samples/Network/kubernetes-require-subscription-vnet/azurepolicy.json]: Same comment about type check - it should come first, before the anyOf

File: samples/Network/kubernetes-require-subscription-vnet/azurepolicy.json [hunk 7]
```diff
@@ -0,0 +1,58 @@
+{
+    "mode": "all",
```
Comment (camillemarie) [inline samples/Network/kubernetes-require-subscription-vnet/azurepolicy.json]: >all [](start = 13, length = 3)

indexed

File: samples/Network/kubernetes-require-subscription-vnet/README.md [hunk 8]
```diff
@@ -0,0 +1,26 @@
+# Ensure that kubernetes Clusters are deployed using the VNet from the subscription
```
Comment (camillemarie) [inline samples/Network/kubernetes-require-subscription-vnet/README.md]: >kubernetes [](start = 14, length = 10)

k=>K

File: samples/Network/kubernetes-require-subscription-vnet/README.md [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+## Try on Portal
+
+[![Deploy to Azure](http://azuredeploy.net/deploybutton.png)](https://portal.azure.com/#blade/Microsoft_Azure_Policy/CreatePolicyDefinitionBlade/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-policy%2Fmaster%2Fsamples%2FNetwork%2Fkubernetes-require-subscription-vnet%2Fazurepolicy.json)
+
+## Try with PowerShell
+
+````powershell
+$definition = New-AzPolicyDefinition -Name "deploy-network-watcher-when-virtual-network-created" -DisplayName "Deploy Kubernetes Cluster with Subscription's VNet" -description "This policy enforces use of subscription's virtual network when creating Kubernetes Clusters, allowing securing network level access to the Kubernetes Cluster." -Policy 'https://raw.githubusercontent.com/Azure/azure-policy/master/samples/Network/kubernetes-require-subscription-vnet/azurepolicy.rules.json' -Parameter 'https://raw.githubusercontent.com/Azure/azure-policy/master/samples/Network/kubernetes-require-subscription-vnet/azurepolicy.parameters.json' -Mode Indexed
```
Comment (camillemarie) [inline samples/Network/kubernetes-require-subscription-vnet/README.md]: >deploy-network-watcher-when-virtual-network-created [](start = 44, length = 51)

kubernetes-require-subscription-vnet

File: samples/Network/hdinsight-require-subscription-vnet/azurepolicy.json (hunk 5 above)
Comment (evsmithmsft) [reply samples/Network/hdinsight-require-subscription-vnet/azurepolicy.json]: This particular code snippet came from Chris, I havent actually seen a case myself, but am following it as best practice.

**Pull request closed without merging**